libtock/hd44780.c

```diff
@@ -0,0 +1,269 @@
+ *
+ * Example:
+ *  hd44780_right_to_left(void);
+ */
+void hd44780_right_to_left(void)
+{
+  int ret = command(DRIVER_HD44780_NUM, 4, 1, 0);
+  if (ret)
+    return;
+}
+
+/* hd44780_autoscroll will 'right justify' the text from the cursor
+ *
+ * Example:
+ *	hd44780_autoscroll(void);
+ */
+void hd44780_autoscroll(void)
+{
+  int ret = command(DRIVER_HD44780_NUM, 5, 0, 0);
+  if (ret)
+    return;
+}
+/* hd44780_no_autoscroll will 'left justify' the text from the cursor
+ *
+ * Example:
+ *	hd44780_no_autoscroll(void);
+ */
+void hd44780_no_autoscroll(void)
+{
+  int ret = command(DRIVER_HD44780_NUM, 5, 1, 0);
+  if (ret)
+    return;
+}
+
+/* hd44780_print_full_string will do some checks for the number of bytes to
+ * be sent and will try to send to the kernel all the bytes in the buffer.
+ * When the syscall returns a certain value (ALLOW_BAD_VALUE) it means that the
+ * command buffer in the kernel is full and a delay is needed in order for the
+ * buffer to be emptied (the commands already saved to be actually executed).
+ * As argument, there is the string to be printed.
+ *
+ * Example:
+ *	hd44780_print_full_string("Here I am!");
+ */
+uint8_t hd44780_print_full_string(char* string)
+{
+  char buff[128];
+  uint8_t len = strlen(string);
+  uint8_t n   = 0;
+  while (n < len) {
+    uint8_t to_write = get_min(sizeof(buff), len - n);
+    memcpy(buff, string, to_write);
+    int ret = allow(DRIVER_HD44780_NUM, 0, (void* )buff, to_write);
+    if (ret == ALLOW_BAD_VALUE) {
+      while (ret == ALLOW_BAD_VALUE) {
+        delay_ms(1000);
+        ret = allow(DRIVER_HD44780_NUM, 0, (void* )buff, to_write);
+      }
+    }
+    n      += ret;
+    string += ret;
+  }
+  return n;
+}
+
+/* hd44780_print_string will do some checks for the number of bytes to be sent
+ * and will try to send to the kernel all the bytes in the buffer.
+ * When the syscall returns a certain value (ALLOW_BAD_VALUE) it stops.
+ * As argument, there is the string to be printed.
+ *
+ * Example:
+ *	hd44780_print_string("Here I am!");
+ */
+uint8_t hd44780_print_string(char* string)
+{
+  char buff[128];
+  uint8_t len = strlen(string);
+  uint8_t n   = 0;
+  while (n < len) {
+    uint8_t to_write = get_min(sizeof(buff), len - n);
+    memcpy(buff, string, to_write);
+    int ret = allow(DRIVER_HD44780_NUM, 0, (void* )buff, to_write);
+    if (!ret || ret == ALLOW_BAD_VALUE) return n;
+    n      += ret;
+    string += ret;
+  }
+  return n;
+}
+
+/* hd44780_set_cursor will set the cursor position to the row and col given as
+ * arguments
+ * As arguments, there are:
+ *	- the column
+ *  - the line
+ *
+ * Example:
+ *	hd44780_set_cursor(5, 1);
+ */
+
+void hd44780_set_cursor(uint8_t col, uint8_t row)
+{
+  int ret = command(DRIVER_HD44780_NUM, 1, col, row);
+  if (ret)
+    return;
+}
+
+/* get_min returns the lower number from the 2 arguments given
+ *
+ */
+uint32_t get_min(uint32_t a, uint32_t b)
```

Comment:
```suggestion
static uint32_t get_min(uint32_t a, uint32_t b)
```

We don't have LTO, so if you can tell the compiler this won't need external linkage it can inline it more aggressively

Reply:
May also need to move this to the top of the file